marxs/source/labSource.py

```diff
@@ -105,3 +107,77 @@ def generate_photons(self, exposuretime):
         photons.add_column(Column(name='dir', data=dir.T))
         photons.add_column(Column(name='polarization', data=polarization_vectors(dir.T, photons['polangle'])))
         return photons
+
+class LabPointSourceCone(Source):
+    '''In-lab point source
+
+    - Photons are uniformly distributed in all directions in cone (from the point). Cone is meant to refer to the volume swept out by a solid angle in a sphere.
+    - Photon start position is source position (center of the sphere = tip of cone)
+
+    Parameters
+    ----------
+    position: 3 element list
+        3D coordinates of photon source
+    direction: array
+        Direciton is given by a vector [x,y,z]
+        This is the direction of the axis along which the beam travels.
+        It is sufficient to enter any vector that spans the beams axis.
+    delta: float 
+        This is half the openning angle of the cone. It is given in steradians.
+    kwargs : see `Source`
+        Other keyword arguments include ``flux``, ``energy`` and ``polarization``.
+        See `Source` for details.
+    '''
+    def __init__(self, position, delta, direction=None,  **kwargs):
+
+        self.dir = direction / np.sqrt(np.dot(direction, direction)) #normalize direction
+        self.position = position
+        self.deltaphi = delta
+        super(LabPointSourceCone, self).__init__(**kwargs)
+
+    def generate_photons(self, exposuretime):
+        photons = super(LabPointSourceCone, self).generate_photons(exposuretime)
+        n = len(photons)
+
+        # assign position to photons
+        pos = np.array([self.position[0] * np.ones(n),
+                        self.position[1] * np.ones(n),
+                        self.position[2] * np.ones(n),
+                        np.ones(n)])
+
+        # randomly choose direction - photons go in all directions from source.
+        #Visualize in arbitrary x-y-z coordinate system (to be reconciled with class parameters later)
+        #Angle from pole (z-axis) = phi. Angle from x-axis is theta
+        #along the pole is the axis of interest i.e. beamline.
+        theta = np.random.uniform(0, 2 * np.pi, n);
+        fractionalArea = 2 * np.pi * (1 - np.cos(self.deltaphi)) / (4 * np.pi) #this is the fractional surface area swept out by delta
+        v = np.random.uniform(0, fractionalArea, n)
+        phi = np.arccos(1 - 2 * v)
+        #for computation of phi see http://www.bogotobogo.com/Algorithms/uniform_distribution_sphere.php
+        
+        dir = np.array([np.cos(theta) * np.sin(phi),
+                        np.sin(theta) * np.sin(phi),
+                        np.cos(phi),
+                        np.zeros(n)])
+
+
```

Comment:
Within a function, use only one empty line to separate blocks. Use two empty lines between classes or between module-level functions.